### ISSUE TITLE
Bug in implicit backward solver for non-leaf tensors

### ISSUE THREAD

**Issue by mh-nguyen712** [deepinv/deepinv#1145]
### Environment

latest

### Describe the bug

The custom backward of least square solvers currently give the incorrect gradient w.r.t the physics parameters, if they are non-leaf tensors. 

### Steps to reproduce

This can happen when, for example we train a neural network to estimate the physics parameters, then do back-propagation though a reconstruction net (eg unrolled, RAM) that uses least square solver:

```python

# %%
import torch
import deepinv as dinv

device = "cuda"
torch.manual_seed(0)

filter = dinv.physics.blur.gaussian_blur(sigma=(4.5, 4.5), device=device)

physics = dinv.physics.Blur(filter=filter, device=device, max_iter=200)
rec_model = dinv.models.RAM(pretrained=True, device=device)
for param in rec_model.parameters():
    param.requires_grad = False
rec_model.eval()


# Train a kernel estimation model
kernel_model = dinv.models.KernelIdentificationNetwork(
    pretrained="download", device=device
)
for p in kernel_model.parameters():
    p.grad = None

x = torch.rand(1, 1, 256, 256, device=device)

# Keep the same measurement for both branches to compare like-for-like gradients.
with torch.no_grad():
    y_fixed = physics(x).detach()


# %%
physics.implicit_backward_solver = False
params = kernel_model(y_fixed)
filter_hat = params["filters"][:, :, 0]
physics.update_parameters(filter=filter_hat)
x_hat = rec_model(y_fixed, physics=physics, sigma=0.01)

physics.filter.retain_grad()  # retain grad for filter to check against implicit backward

loss = (x_hat - x).pow(2).mean()
loss.backward()
true_grad = physics.filter.grad

# %%
physics.implicit_backward_solver = True
physics.filter = physics.filter.detach().clone()
physics.filter.grad = None

for p in kernel_model.parameters():
    p.grad = None
for p in rec_model.parameters():
    p.grad = None
params = kernel_model(y_fixed)
filter_hat = params["filters"][:, :, 0]
physics.update_parameters(filter=filter_hat)
x_hat = rec_model(y_fixed, physics=physics, sigma=0.01)
loss = (x_hat - x).pow(2).mean()
loss.backward()
implicit_grad = physics.filter.grad

# %%
print("True grad:", true_grad)
print("Implicit grad:", implicit_grad)
diff = implicit_grad - true_grad
rel_err = diff.norm() / (true_grad.norm() + 1e-12)
cos = torch.nn.functional.cosine_similarity(
    implicit_grad.flatten(), true_grad.flatten(), dim=0
)
print("Relative error:", rel_err.item())
print("Cosine similarity:", cos.item())

```
This give 

```
True grad: tensor([[[[0.5386, 0.5450, 0.4921,  ..., 0.3816, 0.4019, 0.4210],
          [0.4746, 0.4809, 0.4281,  ..., 0.3166, 0.3370, 0.3563],
          [0.3706, 0.3769, 0.3237,  ..., 0.2099, 0.2306, 0.2502],
          ...,
          [0.3667, 0.3686, 0.3119,  ..., 0.2326, 0.2548, 0.2727],
          [0.4330, 0.4350, 0.3788,  ..., 0.2998, 0.3220, 0.3397],
          [0.4514, 0.4531, 0.3968,  ..., 0.3184, 0.3405, 0.3581]]]],
       device='cuda:0')
Implicit grad: tensor([[[[-3.9124e-02, -1.4905e-03, -2.2849e-02,  ..., -7.5977e-03,
           -9.8136e-03, -8.8548e-03],
          [-3.0722e-02,  6.9987e-03, -1.4152e-02,  ...,  1.5290e-03,
           -8.2322e-04,  2.2074e-05],
          [-4.9971e-02, -1.2088e-02, -3.3668e-02,  ..., -2.0591e-02,
           -2.2791e-02, -2.1940e-02],
          ...,
          [-3.4220e-02,  8.3899e-04, -2.4114e-02,  ..., -7.9989e-03,
           -9.5070e-03, -1.1387e-02],
          [-1.3807e-02,  2.1234e-02, -3.3321e-03,  ...,  1.1570e-02,
            1.0053e-02,  8.1890e-03],
          [-3.9485e-02, -4.9428e-03, -2.9637e-02,  ..., -1.2933e-02,
           -1.4372e-02, -1.6157e-02]]]], device='cuda:0')
Relative error: 1.0545765161514282
Cosine similarity: -0.24902009963989258
```

This is due to a wrong manipulation of grad of non-leaf tensors. 

### Expected behavior

The `implicit_grad` and `true_grad` should be close
